### DISCUSSION TITLE
Kinematic vs Mortar Thermal-Mechanical Contact

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#26879] [Q&A Modules: Solid mechanics]
I am trying to benchmark using kinematic vs mortar for a mechanical contact in a coupled TM model. It is my understanding that the `GapHeatTransfer` function does not do well with penetration of the surfaces, so I am trying kinematic vs mortar for the mechanical contact. 

When using the kinematic contact, the solution progresses well and doesn't seem to have an issue converging. When simply changing the `formulation = kinematic` to `formulation = mortar` the solve does not even solve the first nonlinear iteration. The linear iteration residuals drop by only a small amount and always reach the max number of iterations.

Is there more that needs to be changed when switching to the mortar contact formulation? This is also 3-D if that matters.

Thanks!

**Comment by GiudGiud** [comment]
@bwspenc @jiangwen84

**Comment by lindsayad** [comment]
3D mortar has been exercised significantly less than 2D mortar. Could you share an input, whether publicly or privately? It would make it easier to comment on what the issue might potentially be

**Reply by ghost** [reply]
[contact.zip](https://github.com/idaholab/moose/files/14408725/contact.zip)
I have attached a version of the input file...when you say it hasn't been exercised is that due to something in particular with 3d and mortar, its not the right thing to use or something else? I don't have any particular need to use mortar, its just reading through the documentation and some other conversations lead me to believe it is the "better" formulation especially for TM contact.

**Reply by lindsayad** [reply]
3D mortar has been exercised less mostly because it was added later since the needs at the time for BISON were primarily two-dimensional and the mortaring is more difficult to implement in 3D.

In theory 3D mortar should be a great choice, but there are definitely complexities. The default mortar treatment uses Lagrange multipliers which introduces zero-diagonals in the matrix. Algebraic multigrid (AMG) can't handle such zero-diagonals well. Looking at your first nonlinear iteration, the linear residuals appear to be dropping pretty slowly, which is consistent with AMG not doing well. If I use LU the linear solves (and the nonlinear solve) looks good but it is very slow since LU does not scale well into the hundreds of thousand dofs.

The mechanics folks have recently worked on penalty mortar as well as augmented lagrange mortar. These techniques do not introduce the zero-diagonals like Lagrange multipliers do and are much better candidates for AMG. [This input](https://github.com/idaholab/moose/blob/next/modules/contact/test/tests/pdass_problems/cylinder_friction_penalty_frictional_al_action.i) created by @recuero demonstrates how to setup augmented Lagrange contact. Sadly @recuero has left us, but @dschwen added the augmented Lagrange capability in https://github.com/idaholab/moose/pull/24349 so he should be another good resource

**Reply by ghost** [reply]
Thanks for the background on this. One question with the penalty mortar and the augmented...do they both still provide the zero penetration of kinematic/mortar? My experience with using the gap heat transfer BCs seem to really struggle with penetration that penalty would allow. 

Yeah, the LU is not going to be a workable solution given the size of the problems I would like to scale this to.

**Reply by lindsayad** [reply]
The example Augmented lagrange input I linked to before uses LU. [This input](https://github.com/idaholab/moose/blob/next/modules/contact/test/tests/pdass_problems/cylinder_friction_penalty_frictional_al_action_amg.i) uses Augmented Lagrange with algebraic multigrid. So that would be a better example to look at.  Augmented Lagrange does allow non-zero penetration, however, you can control the allowed penetration using the `Contact/al_penetration_tolerance` parameter. 

We did work on a [Variable Condensation Preconditioner (VCP)](https://mooseframework.inl.gov/source/preconditioners/VariableCondensationPreconditioner.html) that was supposed to allow use of iterative preconditioning (like algebraic multigrid) with Lagrange multipliers (perfect constraint enforcement). However, I don't see any tests in the contact module that employ that preconditioner with algebraic multigrid for frictional contact (although I see one for frictionless [here](https://github.com/idaholab/moose/blob/next/modules/contact/test/tests/dual_mortar/dm_mechanical_contact_precon.i)). Maybe @dewenyushu can comment on the potential of using VCP with frictional contact

**Reply by dewenyushu** [reply]
Thanks @lindsayad for bringing me in the discussion. 

I had VCP working for a multi-physics problem (in Bison) with a considerable size that involves frictional mortar contact in 3D. It showed pretty significant computational benefits back then. However, we could not get AMG to work well with VCP on frictional mortar contact. We still needed to use LU under VCP.

**Reply by lindsayad** [reply]
Any hypotheses why AMG wasn't working?

I also think it would be interesting to explore Schur-complement based field splits for solving LM-based contact, with the LM dofs in contact isolated in the Schur complement block and all other dofs including displacements and non-contact LM dofs in the non-Schur complement split. I think it has the potential of being very effective, but it would take some implementation work to make it happen. We would need to incorporate the LMs and gap into the MOOSE PETSc DM for determining the field decomposition. We would also need to figure out the optimal preconditioner for the Schur complement. For Stokes flow, which is also symmetric, the pressure mass matrix is spectrally equivalent, but in that case the off-diagonal blocks are formed from $\nabla p$ and $\nabla \cdot \vec{v}$ whereas in contact the off-diagonal blocks are simply $\pm \vec{\lambda}$ and $\pm \vec{d}$

**Reply by recuero** [reply]
> Any hypotheses why AMG wasn't working?

It is reported that after condensing the LMs, the system matrix is not necessarily friendly to AMG preconditioners (https://onlinelibrary.wiley.com/doi/full/10.1002/nme.6680). That is, that condensing out the LMs does not make the system "contact property-free" in a way.
